src/java/org/apache/cassandra/tcm/log/LocalLog.java

```diff
@@ -561,15 +586,65 @@ private void notifyPostCommit(ClusterMetadata before, ClusterMetadata after, boo
             ScheduledExecutors.optionalTasks.submit(() -> listener.notifyPostCommit(before, after, fromSnapshot));
     }
 
+    /**
+     * Essentially same as `ready` but throws an unchecked exception
+     */
+    @VisibleForTesting
+    public final ClusterMetadata readyForTests()
+    {
+        try
+        {
+            return ready();
+        }
+        catch (StartupException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public ClusterMetadata ready() throws StartupException
+    {
+        ClusterMetadata metadata = replayPersisted();
+        spec.afterReplay.accept(metadata);
+        logger.info("Marking LocalLog ready at epoch {}", metadata.epoch);
+
+        // Make sure SchemaListener is added if keyspace instance initialization is requested
+        if (spec.initializeKeyspaceInstances && !spec.defaultListeners)
+            throw new IllegalStateException("Keyspace initialization requested, but default listeners are not added.");
+        if (!spec.initializeKeyspaceInstances && spec.defaultListeners)
+            throw new IllegalStateException("Keyspace initialization is disabled, but default listeners are used: they will initialize the keyspace.");
+
+        if (!replayComplete.compareAndSet(false, true))
+            throw new IllegalStateException("Log is already fully initialised");
+
+        logger.debug("Marking LocalLog ready at epoch {}", committed.get().epoch);
+        if (spec.defaultListeners)
+        {
+            logger.info("Adding default listeners to LocalLog");
+            addListeners();
+        }
+        else
+        {
+            logger.info("Adding specified listeners to LocalLog");
+            spec.listeners.forEach(this::addListener);
+            spec.changeListeners.forEach(this::addListener);
+            spec.asyncChangeListeners.forEach(this::addListener);
+        }
+
+        logger.info("Notifying all registered listeners of both pre and post commit event");
+        notifyListeners(spec.initial);
```

Comment:
in the LocalLog constructor we set committed to spec.initial - so here I think we always notifyListeners with the same prev/next metadatas - at least in the reset case we should always notify from an empty cluster metadata I think?

looks like we handle this for schema with that new `isFirst` boolean but I think we need it for all listeners (say we change the state of a nodeid from registered to joined - it wouldn't get properly added to gossip (I think?))